orchagent/pfcactionhandler.cpp

```diff
@@ -0,0 +1,466 @@
+        return;
+    }
+
+    sai_buffer_profile_threshold_mode_t threshold_mode = static_cast<sai_buffer_profile_threshold_mode_t>(attr.value.u32);
+#endif
+
+    // XXX: Remove when above is fixed
+    sai_buffer_profile_threshold_mode_t threshold_mode = SAI_BUFFER_PROFILE_THRESHOLD_MODE_DYNAMIC;
+
+    sai_object_id_t zeroBufferProfileId = SAI_NULL_OBJECT_ID;
+    if (threshold_mode == SAI_BUFFER_PROFILE_THRESHOLD_MODE_STATIC)
+    {
+        zeroBufferProfileId = ZeroBufferProfile::getStaticProfile();
+    }
+    else if (threshold_mode == SAI_BUFFER_PROFILE_THRESHOLD_MODE_DYNAMIC)
+    {
+        zeroBufferProfileId = ZeroBufferProfile::getDynamicProfile();
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Buffer mode in buffer 0x%lx is not supported", queue);
+        return;
+    }
+
+    attr.id = SAI_QUEUE_ATTR_BUFFER_PROFILE_ID;
+    attr.value.oid = zeroBufferProfileId;
+
+    // Set our zero buffer profile
+    status = sai_queue_api->set_queue_attribute(queue, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set buffer profile ID on queue 0x%lx: %d", queue, status);
+        return;
+    }
+
+    // Save original buffer profile
+    m_originalBufferProfile = oldProfileId;
+}
+
+PfcWdZeroBufferHandler::~PfcWdZeroBufferHandler(void)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    attr.id = SAI_QUEUE_ATTR_BUFFER_PROFILE_ID;
+    attr.value.oid = m_originalBufferProfile;
+
+    // Set our zero buffer profile
+    sai_status_t status = sai_queue_api->set_queue_attribute(getQueue(), &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set buffer profile ID on queue 0x%lx: %d", getQueue(), status);
+        return;
+    }
+}
+
+PfcWdZeroBufferHandler::ZeroBufferProfile::ZeroBufferProfile(void)
+{
+    SWSS_LOG_ENTER();
+}
+
+PfcWdZeroBufferHandler::ZeroBufferProfile::~ZeroBufferProfile(void)
+{
+    SWSS_LOG_ENTER();
+
+    destroyStaticProfile();
+    destroyDynamicProfile();
+}
+
+PfcWdZeroBufferHandler::ZeroBufferProfile &PfcWdZeroBufferHandler::ZeroBufferProfile::getInstance(void)
+{
+    SWSS_LOG_ENTER();
+
+    static ZeroBufferProfile instance;
+
+    return instance;
+}
+
+sai_object_id_t PfcWdZeroBufferHandler::ZeroBufferProfile::getStaticProfile(void)
+{
+    SWSS_LOG_ENTER();
+
+    if (getInstance().m_zeroStaticBufferProfile == SAI_NULL_OBJECT_ID)
+    {
+        getInstance().createStaticProfile();
+    }
+
+    return getInstance().m_zeroStaticBufferProfile;
+}
+
+sai_object_id_t PfcWdZeroBufferHandler::ZeroBufferProfile::getDynamicProfile(void)
+{
+    SWSS_LOG_ENTER();
+
+    if (getInstance().m_zeroDynamicBufferProfile == SAI_NULL_OBJECT_ID)
+    {
+        getInstance().createDynamicProfile();
+    }
+
+    return getInstance().m_zeroDynamicBufferProfile;
+}
+
+void PfcWdZeroBufferHandler::ZeroBufferProfile::createStaticProfile(void)
+{
+    SWSS_LOG_ENTER();
+
```

Comment:
Can we pre-create one zero buffer profile and use it for all the ports instead of creating each one dynamically? Do we need to create zero buffer pool? Can we just use the existing pool?

Reply:
Actually, it is exactly the way it works. Class that contains a zero buffer profile is a singleton and creates only one instance of buffer upon a first call.

Reply:
As for pool, I also thought about that. I'll change it in P2 or earlier if I'll have time.